packages/flutter/lib/src/widgets/widget_span.dart

```diff
@@ -0,0 +1,254 @@
+  /// decoration, foreground, background, and spacing options will be used.
+  const WidgetSpan({
+    @required this.child,
+    InlineWidgetAlignment alignment = InlineWidgetAlignment.bottom,
+    TextBaseline baseline,
+    TextStyle style,
+  }) : assert(child != null),
+       assert((alignment == InlineWidgetAlignment.aboveBaseline ||
+               alignment == InlineWidgetAlignment.belowBaseline ||
+               alignment == InlineWidgetAlignment.baseline) ? baseline != null : true),
+       super(alignment:
+           // Convert InlineWidgetAlignment to PlaceholderAlignment in a const fashion.
+           alignment == InlineWidgetAlignment.baseline ? ui.PlaceholderAlignment.baseline :
+           alignment == InlineWidgetAlignment.aboveBaseline ? ui.PlaceholderAlignment.aboveBaseline :
+           alignment == InlineWidgetAlignment.belowBaseline ? ui.PlaceholderAlignment.belowBaseline :
+           alignment == InlineWidgetAlignment.top ? ui.PlaceholderAlignment.top :
+           alignment == InlineWidgetAlignment.bottom ? ui.PlaceholderAlignment.bottom :
+           alignment == InlineWidgetAlignment.middle ? ui.PlaceholderAlignment.middle : null,
+         baseline: baseline, style: style,
+       );
+
+  /// The widget to embed inline within text.
+  final Widget child;
+
+  /// Adds a placeholder box to the paragraph builder if a size has been
+  /// calculated for the widget.
+  ///
+  /// Sizes are provided through [dimensions], which should contain a 1:1
+  /// in-order mapping of widget to laid out dimensions. If no such dimension
+  /// is provided, the widget will be skipped.
+  ///
+  /// Since widget sizes are calculated independently from the rest of the
+  /// paragraph, the [textScaleFactor] is ignored.
+  @override
+  void build(ui.ParagraphBuilder builder, { double textScaleFactor = 1.0, List<PlaceholderDimensions> dimensions }) {
+    assert(debugAssertIsValid());
+
+    final bool hasStyle = style != null;
+    if (hasStyle)
+      builder.pushStyle(style.getTextStyle(textScaleFactor: textScaleFactor));
+    if (dimensions != null) {
+      assert(builder.placeholderCount < dimensions.length);
+      final PlaceholderDimensions currentDimensions = dimensions[builder.placeholderCount];
+      builder.addPlaceholder(
+        currentDimensions.size.width,
+        currentDimensions.size.height,
+        alignment,
+        baseline: currentDimensions.baseline,
+        baselineOffset: currentDimensions.baselineOffset,
+      );
+    }
+    if (hasStyle)
+      builder.pop();
+  }
+
+  /// Calls visitor on this [WidgetSpan]. There are no children spans to walk.
+  @override
+  bool visitChildren(InlineSpanVisitor visitor) {
+    if (!visitor(this))
+      return false;
+    return true;
+  }
+
+  @override
+  InlineSpan getSpanForPositionVisitor(TextPosition position, TrackingInt offset) {
+    return null;
+  }
+
+  @override
+  int codeUnitAt(int index) {
+    return null;
+  }
```

Comment:
some of these don't need to be implemented here if the superclass has the default implementation that works in the general case